sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+                kDifPadctrlChangeOk);
+    }
+  }
+
+  /**
+   * Tests a failure to enable an attribute due to a conflict
+   *
+   * @param attr_field The attribute field to test
+   * @param pad_kind What kind of pad `attr_field` is for
+   */
+  void TestEnableAttrFailureConflict(const AttrField &attr_field,
+                                     dif_padctrl_pad_kind_t pad_kind) {
+    DeviceInit();
+
+    ExpectLockCheck(false);
+
+    // Expect a read of `attr_field` with a couple of attributes set.
+    ExpectAttrFieldRead(attr_field, dif_padctrl_attr_t(kDifPadctrlAttrIOInvert |
+                                                       kDifPadctrlAttrPullUp));
+
+    // Attempt to set an attribute that conflicts with the currently set ones
+    // expecting the appropriate error.
+    EXPECT_EQ(
+        dif_padctrl_enable_attrs(&dif_padctrl_, pad_kind, attr_field.pad_num,
+                                 kDifPadctrlAttrPullDown),
+        kDifPadctrlChangeAttrConflict);
+  }
+
+  /**
+   * Tests a failure to write attributes due to a conflict
+   *
+   * @param attr_field The attribute field to test
+   * @param pad_kind What kind of pad `attr_field` is for
+   */
+  void TestWriteAttrFailureConflict(const AttrField &attr_field,
+                                    dif_padctrl_pad_kind_t pad_kind) {
+    DeviceInit();
+
+    ExpectLockCheck(false);
+
+    // Expect a read of `attr_field` with an attribute set.
+    ExpectAttrFieldRead(attr_field,
+                        dif_padctrl_attr_t(kDifPadctrlAttrIOInvert));
+
+    // Write back attributes that conflict with each other and not the attribute
+    // that was initially set expecting the appropriate error.
+    dif_padctrl_attr_t conflicting_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrPullUp | kDifPadctrlAttrPullDown);
+
+    EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind,
+                                      attr_field.pad_num, conflicting_attrs),
+              kDifPadctrlChangeAttrConflict);
+  }
+
+  dif_padctrl_t dif_padctrl_;
+};
+
+TEST_F(DifPadctrlTest, InitSuccess) { DeviceInit(); }
+
+TEST_F(DifPadctrlTest, InitFailureInvalidPadctrl) {
+  mmio_region_t base_addr = dev().region();
+
+  EXPECT_EQ(dif_padctrl_init(base_addr, nullptr), kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, LockSuccess) {
+  DeviceInit();
+
+  EXPECT_WRITE32(PADCTRL_REGEN_REG_OFFSET, 0);
+
+  EXPECT_EQ(dif_padctrl_lock(&dif_padctrl_), kDifPadctrlOk);
+}
+
+TEST_F(DifPadctrlTest, LockFailureInvalidPadctrl) {
+  EXPECT_EQ(dif_padctrl_lock(nullptr), kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, ResetSuccess) {
+  dif_padctrl_attr_t reset_attrs =
+      dif_padctrl_attr_t(kDifPadctrlAttrKeeper | kDifPadctrlAttrIOInvert);
+
+  uint32_t replicated_attrs = reset_attrs |
+                              (reset_attrs << PADCTRL_PARAM_ATTRDW) |
+                              (reset_attrs << PADCTRL_PARAM_ATTRDW * 2) |
+                              (reset_attrs << PADCTRL_PARAM_ATTRDW * 3);
```

Comment:
Prefer to move variables to just before they are initialized. For example, this code would be better written like this:
```c
DeviceInit();
ExpectLockCheck(false);

dif_padctrl_attr_t reset_attrs = ...;
uint32_t replicated_attrs = ...;
ExpectReset(replicated_attrs);
EXPECT_EQ(...);
```

This comment applies to all other tests below that initialize some value to immediately use in exactly one followup statement.

Reply:
I've gone through variable declarations and rearranged things.